python_modules/dagster/dagster/_components/__init__.py

```diff
@@ -0,0 +1,179 @@
+import importlib.util
+import os
+import sys
+from abc import ABC, abstractmethod
+from types import ModuleType
+from typing import TYPE_CHECKING, ClassVar, Dict, Final, Iterable, Optional, Type
+
+from dagster._core.errors import DagsterError
+from dagster._utils import to_snake_case
+
+if TYPE_CHECKING:
+    from dagster._core.definitions.definitions_class import Definitions
+
+
+class Component(ABC):
+    name: ClassVar[Optional[str]] = None
+
+    @classmethod
+    def registered_name(cls):
+        return cls.name or to_snake_case(cls.__name__)
+
+    @classmethod
+    def generate_files(cls) -> None:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def build_defs(self) -> "Definitions": ...
+
+
+def is_inside_deployment_project(path: str = ".") -> bool:
+    try:
+        _resolve_deployment_root(path)
+        return True
+    except DagsterError:
+        return False
+
+
+def _resolve_deployment_root(path: str) -> str:
+    current_path = os.path.abspath(path)
+    while not _is_deployment_root(current_path):
+        current_path = os.path.dirname(current_path)
+        if current_path == "/":
+            raise DagsterError("Cannot find deployment root")
+    return current_path
+
+
+def is_inside_code_location_project(path: str = ".") -> bool:
+    try:
+        _resolve_code_location_root(path)
+        return True
+    except DagsterError:
+        return False
+
+
+def _resolve_code_location_root(path: str) -> str:
+    current_path = os.path.abspath(path)
+    while not _is_code_location_root(current_path):
+        current_path = os.path.dirname(current_path)
+        if current_path == "/":
+            raise DagsterError("Cannot find code location root")
+    return current_path
+
+
+def _is_deployment_root(path: str) -> bool:
```

Comment:
there should be a more deterministic/bulletproof way to detect this

Reply:
Yes, this is a temporary implementation. When the project structure is more nailed down (after say we start using `uv`) I can change it to something more robust.